[build] Cannot build raylib 2.0.0-dev with -DSHARE on FreeBSD

Hello i have problem with compilation on FreeBSD 11 with raylib 2.0.0-dev.

```
#!/bin/sh
mkdir build
cd build
cmake .. -DCMAKE_BUILD_TYPE=debug -DSHARED=ON -DBUILD_EXAMPLES=NO -DBUILD_GAMES=NO
gmake -j3
cd ..
```


```
-- The C compiler identification is Clang 4.0.0
-- The CXX compiler identification is Clang 4.0.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Performing Test COMPILER_HAS_THOSE_TOGGLES
-- Performing Test COMPILER_HAS_THOSE_TOGGLES - Success
-- Testing if -Werror=pointer-arith can be used -- works
-- Testing if -Werror=implicit-function-declaration can be used -- works
-- Testing if -fno-strict-aliasing can be used -- works
-- Testing if file system supports symlinks
-- Testing if file system supports symlinks -- supported
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Could NOT find Vulkan (missing: VULKAN_LIBRARY VULKAN_INCLUDE_DIR) 
-- Using X11 for window creation
-- Looking for XOpenDisplay in /usr/local/lib/libX11.so;/usr/local/lib/libXext.so
-- Looking for XOpenDisplay in /usr/local/lib/libX11.so;/usr/local/lib/libXext.so - found
-- Looking for gethostbyname
-- Looking for gethostbyname - found
-- Looking for connect
-- Looking for connect - found
-- Looking for remove
-- Looking for remove - found
-- Looking for shmat
-- Looking for shmat - found
-- Looking for IceConnectionNumber in ICE
-- Looking for IceConnectionNumber in ICE - found
-- Found X11: /usr/local/lib/libX11.so
-- Compiling with the flags:
--   PLATFORM=PLATFORM_DESKTOP
--   GRAPHICS=GRAPHICS_API_OPENGL_33
-- Configuring done
-- Generating done
-- Build files have been written to: /usr/home/maxfx/Documents/raylib/build
Scanning dependencies of target glfw
[  2%] Building C object release/external/glfw/src/CMakeFiles/glfw.dir/init.c.o
[  7%] Building C object release/external/glfw/src/CMakeFiles/glfw.dir/context.c.o
[  7%] Building C object release/external/glfw/src/CMakeFiles/glfw.dir/input.c.o
[ 10%] Building C object release/external/glfw/src/CMakeFiles/glfw.dir/monitor.c.o
[ 12%] Building C object release/external/glfw/src/CMakeFiles/glfw.dir/vulkan.c.o
[ 15%] Building C object release/external/glfw/src/CMakeFiles/glfw.dir/window.c.o
[ 17%] Building C object release/external/glfw/src/CMakeFiles/glfw.dir/x11_init.c.o
[ 20%] Building C object release/external/glfw/src/CMakeFiles/glfw.dir/x11_monitor.c.o
[ 23%] Building C object release/external/glfw/src/CMakeFiles/glfw.dir/x11_window.c.o
[ 25%] Building C object release/external/glfw/src/CMakeFiles/glfw.dir/xkb_unicode.c.o
[ 28%] Building C object release/external/glfw/src/CMakeFiles/glfw.dir/posix_time.c.o
[ 30%] Building C object release/external/glfw/src/CMakeFiles/glfw.dir/posix_thread.c.o
[ 33%] Building C object release/external/glfw/src/CMakeFiles/glfw.dir/glx_context.c.o
[ 35%] Building C object release/external/glfw/src/CMakeFiles/glfw.dir/egl_context.c.o
[ 38%] Building C object release/external/glfw/src/CMakeFiles/glfw.dir/osmesa_context.c.o
[ 41%] Building C object release/external/glfw/src/CMakeFiles/glfw.dir/null_joystick.c.o
[ 43%] Linking C static library libglfw3.a
[ 43%] Built target glfw
Scanning dependencies of target raylib
Scanning dependencies of target raylib_shared
[ 48%] Building C object release/CMakeFiles/raylib_shared.dir/audio.c.o
[ 48%] Building C object release/CMakeFiles/raylib.dir/audio.c.o
[ 51%] Building C object release/CMakeFiles/raylib.dir/core.c.o
[ 53%] Building C object release/CMakeFiles/raylib_shared.dir/core.c.o
[ 56%] Building C object release/CMakeFiles/raylib.dir/models.c.o
[ 58%] Building C object release/CMakeFiles/raylib.dir/rlgl.c.o
[ 61%] Building C object release/CMakeFiles/raylib.dir/shapes.c.o
[ 64%] Building C object release/CMakeFiles/raylib_shared.dir/models.c.o
[ 66%] Building C object release/CMakeFiles/raylib_shared.dir/rlgl.c.o
[ 69%] Building C object release/CMakeFiles/raylib.dir/text.c.o
[ 71%] Building C object release/CMakeFiles/raylib.dir/textures.c.o
[ 74%] Building C object release/CMakeFiles/raylib_shared.dir/shapes.c.o
[ 76%] Building C object release/CMakeFiles/raylib.dir/utils.c.o
[ 79%] Building C object release/CMakeFiles/raylib_shared.dir/text.c.o
[ 82%] Building C object release/CMakeFiles/raylib.dir/external/mini_al.c.o
[ 84%] Building C object release/CMakeFiles/raylib_shared.dir/textures.c.o
[ 87%] Building C object release/CMakeFiles/raylib_shared.dir/utils.c.o
[ 89%] Building C object release/CMakeFiles/raylib.dir/external/stb_vorbis.c.o
[ 92%] Building C object release/CMakeFiles/raylib_shared.dir/external/mini_al.c.o
[ 94%] Linking C static library libraylib.a
[ 94%] Built target raylib
[ 97%] Building C object release/CMakeFiles/raylib_shared.dir/external/stb_vorbis.c.o
[100%] Linking C shared library libraylib.so
/usr/bin/ld: cannot find -lX11
cc: error: linker command failed with exit code 1 (use -v to see invocation)
gmake[2]: *** [release/CMakeFiles/raylib_shared.dir/build.make:335: release/libraylib.so.2.0.0] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:127: release/CMakeFiles/raylib_shared.dir/all] Error 2
gmake: *** [Makefile:152: all] Error 2

```

--- a/.travis.yml
+++ b/.travis.yml
@@ -96,6 +96,7 @@ script:
   - if [[ "$ARCH" != *-android && "$ARCH" != html5 ]]; then
       pkg-config --static --libs raylib;
       nm -g release/libraylib.a | grep glfwGetProcAddress || (echo "libraylib.a doesn't contain GLFW symbols! Aborting..." && false);
+      ctest --output-on-failure;
     fi
 
 deploy:

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -50,3 +50,5 @@ endif()
 if (${BUILD_GAMES})
   add_subdirectory(games)
 endif()
+
+enable_testing()

--- a/cmake/test-pkgconfig.sh
+++ b/cmake/test-pkgconfig.sh
@@ -1,0 +1,21 @@
+#!/bin/sh
+# Test if including/linking/running an installed raylib works
+
+set -x
+export LD_RUN_PATH=/usr/local/lib
+
+CFLAGS="-Wall -Wextra -Werror $CFLAGS"
+if [ "$ARCH" = "i386" ]; then
+CFLAGS="-m32 $CLFAGS"
+fi
+
+cat << EOF | ${CC:-cc} -otest -xc - $(pkg-config --libs --cflags $@ raylib.pc) $CFLAGS && exec ./test
+#include <stdlib.h>
+#include <raylib.h>
+
+int main(void)
+{
+    int num = GetRandomValue(42, 1337);
+    return 42 <= num && num <= 1337 ? EXIT_SUCCESS : EXIT_FAILURE;
+}
+EOF

--- a/cmake/utils.cmake
+++ b/cmake/utils.cmake
@@ -11,13 +11,8 @@ if(${PLATFORM} MATCHES "Android")
 elseif(${PLATFORM} MATCHES "Web")
 elseif(APPLE)
   find_library(OPENGL_LIBRARY OpenGL)
-  find_library(COCOA_LIBRARY Cocoa)
-  find_library(IOKIT_LIBRARY IOKit)
-  find_library(COREFOUNDATION_LIBRARY CoreFoundation)
-  find_library(COREVIDEO_LIBRARY CoreVideo)
 
-  set(LIBS_PRIVATE ${OPENGL_LIBRARY} ${COCOA_LIBRARY}
-                   ${IOKIT_LIBRARY} ${COREFOUNDATION_LIBRARY} ${COREVIDEO_LIBRARY})
+  set(LIBS_PRIVATE ${OPENGL_LIBRARY})
 elseif(WIN32)
   # no pkg-config --static on Windows yet...
 else()
@@ -34,12 +29,6 @@ else()
   endif()
 
   set(LIBS_PRIVATE m pthread ${OPENGL_LIBRARIES} ${OSS_LIBRARY})
-  # TODO: maybe read those out of glfw's cmake config?
-  if(USE_WAYLAND)
-    set(LIBS_PRIVATE ${LIBS_PRIVATE} wayland-client wayland-cursor wayland-egl)
-  else()
-    set(LIBS_PRIVATE ${LIBS_PRIVATE} X11 Xrandr Xinerama Xi Xxf86vm Xcursor)
-  endif()
 endif()
 
 if(${PLATFORM} MATCHES "Desktop")
@@ -53,14 +42,8 @@ if(${PLATFORM} MATCHES "Desktop")
   endif()
 endif()
 
-# Ugly crutch. Temporary workaround for #551
-if("${CMAKE_SYSTEM_NAME}" MATCHES "(Free|Net|Open)BSD|DragonFly")
-    link_directories("${CMAKE_INSTALL_PREFIX}/lib")
-endif()
-
 if(CMAKE_SYSTEM_NAME STREQUAL Linux)
   set(LINUX TRUE)
-  set(LIBS_PRIVATE dl ${LIBS_PRIVATE})
 endif()
 
 foreach(L ${LIBS_PRIVATE})

--- a/raylib.pc.in
+++ b/raylib.pc.in
@@ -9,5 +9,5 @@ URL: http://github.com/raysan5/raylib
 Version: @PROJECT_VERSION@
 Libs: -L${libdir} -lraylib
 Libs.private: @PKG_CONFIG_LIBS_PRIVATE@
-Requires.private:
+Requires.private: @GLFW_PKG_DEPS@
 Cflags: -I${includedir}

--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -144,6 +144,8 @@ if(${SHARED})
       PUBLIC_HEADER DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}"
     )
   endif()
+
+  add_test("pkg-config" ${PROJECT_SOURCE_DIR}/../cmake/test-pkgconfig.sh)
 endif(${SHARED})
 
 if(${STATIC})
@@ -167,6 +169,8 @@ if(${STATIC})
     ARCHIVE DESTINATION "${CMAKE_INSTALL_LIBDIR}"
     PUBLIC_HEADER DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}"
   )
+
+  add_test("pkg-config--static" ${PROJECT_SOURCE_DIR}/../cmake/test-pkgconfig.sh --static)
 endif(${STATIC})
 
 configure_file(../raylib.pc.in raylib.pc @ONLY)
@@ -203,3 +207,5 @@ SET(CPACK_RESOURCE_FILE_LICENSE "${PROJECT_SOURCE_DIR}/../LICENSE.md")
 SET(CPACK_PACKAGE_FILE_NAME "raylib-${PROJECT_VERSION}$ENV{RAYLIB_PACKAGE_SUFFIX}")
 SET(CPACK_GENERATOR "ZIP;TGZ") # Remove this, if you want the NSIS installer on Windows
 include(CPack)
+
+enable_testing()

--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -33,6 +33,8 @@ if(NOT glfw3_FOUND AND NOT USE_EXTERNAL_GLFW STREQUAL "ON" AND "${PLATFORM}" MAT
   include_directories(external/glfw/include)
 
   list(APPEND raylib_sources $<TARGET_OBJECTS:glfw_objlib>)
+else()
+  set(GLFW_PKG_DEPS glfw)
 endif()
 
 include(utils)
@@ -92,7 +94,7 @@ elseif(${PLATFORM} MATCHES "Android")
   add_if_flag_compiles(-Wa,--noexecstack CMAKE_C_FLAGS)
   add_if_flag_compiles(-no-canonical-prefixes CMAKE_C_FLAGS)
   add_definitions(-DANDROID -D__ANDROID_API__=21)
-  include_directories(external/android/native_app_glue )
+  include_directories(external/android/native_app_glue)
   set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -Wl,--exclude-libs,libatomic.a -Wl,--build-id -Wl,-z,noexecstack -Wl,-z,relro -Wl,-z,now -Wl,--warn-shared-textrel -Wl,--fatal-warnings -uANativeActivity_onCreate")
 
 elseif(${PLATFORM} MATCHES "Raspberry Pi")
@@ -114,8 +116,8 @@ if(${SHARED})
   set(CMAKE_MACOSX_RPATH ON)
 
   target_link_libraries(${RAYLIB}_shared ${LIBS_PRIVATE})
-  if (${PLATFORM} MATCHES "Desktop")
-    target_link_libraries(${RAYLIB}_shared glfw ${GLFW_LIBRARIES})
+  if (${PLATFORM} MATCHES "PLATFORM_DESKTOP")
+    target_link_libraries(${RAYLIB}_shared glfw)
   endif()
   if (UNIX AND ${FILESYSTEM_LACKS_SYMLINKS})
     MESSAGE(WARNING "Can't version UNIX shared library on file system without symlink support")
@@ -151,8 +153,9 @@ if(${STATIC})
 
   add_library(${RAYLIB} STATIC ${sources})
 
-  set(PKG_CONFIG_LIBS_PRIVATE ${__PKG_CONFIG_LIBS_PRIVATE})
-  if (${PLATFORM} MATCHES "Desktop")
+  set(PKG_CONFIG_LIBS_PRIVATE ${__PKG_CONFIG_LIBS_PRIVATE} ${GLFW_PKG_LIBS})
+  string (REPLACE ";" " " PKG_CONFIG_LIBS_PRIVATE "${PKG_CONFIG_LIBS_PRIVATE}")
+  if (${PLATFORM} MATCHES "PLATFORM_DESKTOP")
     target_link_libraries(${RAYLIB} glfw ${GLFW_LIBRARIES})
   endif()
 

--- a/src/external/glfw/CMakeLists.txt
+++ b/src/external/glfw/CMakeLists.txt
@@ -327,10 +327,10 @@ endif()
 # Export GLFW library dependencies
 #--------------------------------------------------------------------
 foreach(arg ${glfw_PKG_DEPS})
-    set(GLFW_PKG_DEPS "${GLFW_PKG_DEPS} ${arg}")
+    set(GLFW_PKG_DEPS "${GLFW_PKG_DEPS} ${arg}" PARENT_SCOPE)
 endforeach()
 foreach(arg ${glfw_PKG_LIBS})
-    set(GLFW_PKG_LIBS "${GLFW_PKG_LIBS} ${arg}")
+    set(GLFW_PKG_LIBS "${GLFW_PKG_LIBS} ${arg}" PARENT_SCOPE)
 endforeach()
 
 #--------------------------------------------------------------------

